### PR TITLE
8266330: itableMethodEntry::initialize() asserts with archived old classes

### PR DESCRIPTION
**Pull request by calvinccheung**
```
#ifdef ASSERT 
  if (MetaspaceShared::is_in_shared_metaspace((void*)&_method) && 
     !MetaspaceShared::remapped_readwrite() && 
     !MetaspaceShared::is_old_class(m->method_holder())) { 
    // At runtime initialize_itable is rerun as part of link_class_impl() 
    // for a shared class loaded by the non-boot loader. 
    // The dumptime itable method entry should be the same as the runtime entry. 
    assert(_method == m, "sanity"); 
  } 
#endif
```

The above includes the check for !MetaspaceShared::is_old_class(m->method_holder()).
While the method_holder() is not an old class, the _klass in the klassITable could be an old class.
The proposed fix is to have the caller of the above function pass in the _klass and add another check
making sure the _klass is not an old class before the assert.

Testing:
- [x] tiers 1 and 2 (including the new test)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266330](https://bugs.openjdk.java.net/browse/JDK-8266330): itableMethodEntry::initialize() asserts with archived old classes

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 6b83ea9ff5793897f5e6bcf348e151b940583679
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**) ⚠️ Review applies to 6b83ea9ff5793897f5e6bcf348e151b940583679

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3930/head:pull/3930` \
`$ git checkout pull/3930`

Update a local copy of the PR: \
`$ git checkout pull/3930` \
`$ git pull https://git.openjdk.java.net/jdk pull/3930/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3930`

View PR using the GUI difftool: \
`$ git pr show -t 3930`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3930.diff">https://git.openjdk.java.net/jdk/pull/3930.diff</a>

</details>

### CODE REVIEW

**Comment by coleenp** [comment]
Hi, the concept of 'old' class conflicts with the concept of 'old' methods which were ones that were redefined.  I looked at this to see how redefinition could have broken this. One of these should change their name.
In a future RFE, can you rename is_old_class to has_old_class_version(klass) ?

**Review by iklam — APPROVED** [review]
LGTM

File: src/hotspot/share/oops/klassVtable.cpp [hunk 1]
```diff
@@ -1088,16 +1088,18 @@ void klassVtable::dump_vtable() {
...
 // Initialize a itableMethodEntry
-void itableMethodEntry::initialize(Method* m) {
+void itableMethodEntry::initialize(InstanceKlass* klass, Method* m) {
   if (m == NULL) return;
 
 #ifdef ASSERT
   if (MetaspaceShared::is_in_shared_metaspace((void*)&_method) &&
      !MetaspaceShared::remapped_readwrite() &&
-     !MetaspaceShared::is_old_class(m->method_holder())) {
+     !MetaspaceShared::is_old_class(m->method_holder()) && !MetaspaceShared::is_old_class(klass)) {
```
Comment (yminqi) [inline src/hotspot/share/oops/klassVtable.cpp:1097]: The newly added check should start with a new line as existing checks.

**Review by yminqi — APPROVED** [review]
LGTM, minor nits.

**Comment by calvinccheung** [comment]
> LGTM, minor nits.

@yminqi Thanks for your review. I've fixed the nit.

**Pull request closed without merging**